lib/retry/retry.go

```diff
@@ -0,0 +1,37 @@
+package retry
+
+import (
+	"errors"
+	"reflect"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+)
+
+var log = logging.Logger("retry")
+
+func errorIsIn(err error, errorTypes []error) bool {
+	for _, etype := range errorTypes {
+		tmp := reflect.New(reflect.PointerTo(reflect.ValueOf(etype).Elem().Type())).Interface()
+		if errors.As(err, tmp) {
+			return true
+		}
+	}
+	return false
+}
+
+func Retry[T any](attempts int, sleep int, errorTypes []error, f func() (T, error)) (result T, err error) {
```

Comment:
I'd make the `sleep` param be time.Duration.

Reply:
I'd also call the sleep param something that relays that it's increasing - e.g. `initialBackoff`

Reply:
done